# Random forest 🌲🌲

## **Introduction**

Here, you will train, tune, and evaluate a random forest model using data from spreadsheet of survey responses from 129,880 customers. It includes data points such as class, flight distance, and inflight entertainment. Random forest model will be used to predict whether a customer will be satisfied with their flight experience.

## **Imports 🚢** 

In [25]:
#operational package 
import numpy as np
import pandas as pd
import pickle as pkl

#model & evaluation

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, PredefinedSplit, GridSearchCV
from sklearn.metrics import f1_score, precision_score, recall_score, accuracy_score

In [26]:
#load the data 
air_data = pd.read_csv("Invistico_Airline.csv")

## **Data cleaning 🧹🫧** 

In [27]:
# Display first 10 rows.
air_data.head(10)

satisfaction   Customer Type  Age   Type of Travel     Class  \
0    satisfied  Loyal Customer   65  Personal Travel       Eco   
1    satisfied  Loyal Customer   47  Personal Travel  Business   
2    satisfied  Loyal Customer   15  Personal Travel       Eco   
3    satisfied  Loyal Customer   60  Personal Travel       Eco   
4    satisfied  Loyal Customer   70  Personal Travel       Eco   
5    satisfied  Loyal Customer   30  Personal Travel       Eco   
6    satisfied  Loyal Customer   66  Personal Travel       Eco   
7    satisfied  Loyal Customer   10  Personal Travel       Eco   
8    satisfied  Loyal Customer   56  Personal Travel  Business   
9    satisfied  Loyal Customer   22  Personal Travel       Eco   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0              265             0                                  0   
1             2464             0                                  0   
2             2138             0                                  0   
3              623             0                                  0   
4              354             0                                  0   
5             1894             0                                  0   
6              227             0                                  0   
7             1812             0                                  0   
8               73             0                                  0   
9             1556             0                                  0   

   Food and drink  Gate location  ...  Online support  Ease of Online booking  \
0               0              2  ...               2                       3   
1               0              3  ...               2                       3   
2               0              3  ...               2                       2   
3               0              3  ...               3                       1   
4               0              3  ...               4                       2   
5               0              3  ...               2                       2   
6               0              3  ...               5                       5   
7               0              3  ...               2                       2   
8               0              3  ...               5                       4   
9               0              3  ...               2                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   
5                 5                 4                 5                5   
6                 5                 0                 5                5   
7                 3                 3                 4                5   
8                 4                 0                 1                5   
9                 2                 4                 5                3   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   
5            4                2                           0   
6            5                3                          17   
7            4                2                           0   
8            4                4                           0   
9            4                2                          30   

   Arrival Delay in Minutes  
0                       0.0  
1         

In [28]:
# Display variable names and types.
air_data.dtypes

satisfaction                          object
Customer Type                         object
Age                                    int64
Type of Travel                        object
Class                                 object
Flight Distance                        int64
Seat comfort                           int64
Departure/Arrival time convenient      int64
Food and drink                         int64
Gate location                          int64
Inflight wifi service                  int64
Inflight entertainment                 int64
Online support                         int64
Ease of Online booking                 int64
On-board service                       int64
Leg room service                       int64
Baggage handling                       int64
Checkin service                        int64
Cleanliness                            int64
Online boarding                        int64
Departure Delay in Minutes             int64
Arrival Delay in Minutes             float64
dtype: obj

In [29]:
# number of rows and the number of columns.
air_data.shape

(129880, 22)

In [30]:
# find missing values in data.
air_data.isna().any(axis=1).sum()

393

**Question:** How many rows of data are missing values?**

There are 393 rows with missing values.

In [31]:
# Drop missing values.
air_data_subset = air_data.dropna(axis=0)

# Display the first 10 rows.
air_data_subset.tail(20)

satisfaction      Customer Type  Age   Type of Travel     Class  \
129859  dissatisfied  disloyal Customer   29  Personal Travel       Eco   
129860  dissatisfied  disloyal Customer   43  Personal Travel       Eco   
129861     satisfied  disloyal Customer   36  Personal Travel       Eco   
129862     satisfied  disloyal Customer   25  Personal Travel       Eco   
129863     satisfied  disloyal Customer   52  Personal Travel       Eco   
129864     satisfied  disloyal Customer   24  Personal Travel       Eco   
129865     satisfied  disloyal Customer   59  Personal Travel       Eco   
129866     satisfied  disloyal Customer   18  Personal Travel       Eco   
129867     satisfied  disloyal Customer   30  Personal Travel       Eco   
129868     satisfied  disloyal Customer   45  Personal Travel       Eco   
129869     satisfied  disloyal Customer   55  Personal Travel       Eco   
129870     satisfied  disloyal Customer   70  Personal Travel       Eco   
129871     satisfied  disloyal Customer   35  Personal Travel       Eco   
129872     satisfied  disloyal Customer   69  Personal Travel       Eco   
129874     satisfied  disloyal Customer   11  Personal Travel       Eco   
129875     satisfied  disloyal Customer   29  Personal Travel       Eco   
129876  dissatisfied  disloyal Customer   63  Personal Travel  Business   
129877  dissatisfied  disloyal Customer   69  Personal Travel       Eco   
129878  dissatisfied  disloyal Customer   66  Personal Travel       Eco   
129879  dissatisfied  disloyal Customer   38  Personal Travel       Eco   

        Flight Distance  Seat comfort  Departure/Arrival time convenient  \
129859             2370             4                                  5   
129860             2748             4                                  5   
129861             3466             0                                  4   
129862             2220             4                                  4   
129863             1602             4                                  5   
129864             2039             4                                  5   
129865             2641             4                                  5   
129866             1243             5                                  0   
129867             1961             5                                  1   
129868             1612             5                                  2   
129869             1953             5                                  2   
129870             1674             5                                  4   
129871             3287             5                                  4   
129872             2240             5                                  4   
129874             2752             5                                  5   
129875             1731             5                                  5   
129876             2087             2                                  3   
129877             2320             3                                  0   
129878             2450             3                                  2   
129879             4307             3                                  4   

        Food and drink  Gate location  ...  Online support  \
129859               4              3  ...               4   
129860               5              1  ...               5   
129861               0              3  ...               4   
129862               3              4  ...               4   
129863               4              2  ...               3   
129864               4              2  ...               2   
129865               4              3  ...               2   
129866               5              1  ...               2   
129867               5              4  ...               5   
129868               5              3  ...               2   
129869               5              4  ...               5   
129870               5              1  ...               5   
129871               5       

In [32]:
# Count of missing values.
air_data_subset.isna().sum()

satisfaction                         0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Seat comfort                         0
Departure/Arrival time convenient    0
Food and drink                       0
Gate location                        0
Inflight wifi service                0
Inflight entertainment               0
Online support                       0
Ease of Online booking               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Cleanliness                          0
Online boarding                      0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
dtype: int64

In [33]:
# Convert categorical features to one-hot encoded features.
air_data_subset_dummies = pd.get_dummies(air_data_subset, 
                                         columns=['Customer Type','Type of Travel','Class'])

# Display the first 10 rows.
air_data_subset_dummies.head(10)

satisfaction  Age  Flight Distance  Seat comfort  \
0    satisfied   65              265             0   
1    satisfied   47             2464             0   
2    satisfied   15             2138             0   
3    satisfied   60              623             0   
4    satisfied   70              354             0   
5    satisfied   30             1894             0   
6    satisfied   66              227             0   
7    satisfied   10             1812             0   
8    satisfied   56               73             0   
9    satisfied   22             1556             0   

   Departure/Arrival time convenient  Food and drink  Gate location  \
0                                  0               0              2   
1                                  0               0              3   
2                                  0               0              3   
3                                  0               0              3   
4                                  0               0              3   
5                                  0               0              3   
6                                  0               0              3   
7                                  0               0              3   
8                                  0               0              3   
9                                  0               0              3   

   Inflight wifi service  Inflight entertainment  Online support  ...  \
0                      2                       4               2  ...   
1                      0                       2               2  ...   
2                      2                       0               2  ...   
3                      3                       4               3  ...   
4                      4                       3               4  ...   
5                      2                       0               2  ...   
6                      2                       5               5  ...   
7                      2                       0               2  ...   
8                      5                       3               5  ...   
9                      2                       0               2  ...   

   Online boarding  Departure Delay in Minutes  Arrival Delay in Minutes  \
0                2                           0                       0.0   
1                2                         310                     305.0   
2                2                           0                       0.0   
3                3                           0                       0.0   
4                5                           0                       0.0   
5                2                           0                       0.0   
6                3                          17                      15.0   
7                2                           0                       0.0   
8                4                           0                       0.0   
9                2                          30                      26.0   

   Customer Type_Loyal Customer  Customer Type_disloyal Customer  \
0                             1                                0   
1                             1                                0   
2                             1                                0   
3                             1                                0   
4                             1                                0   
5                             1                                0   
6                             1                                0   
7                             1                                0   
8                             1                                0   
9                             1                                0   

   Type of Travel_Business travel  Type of Travel_Personal Travel  \
0                               0                               1   
1                               0                               1   
2                           

In [34]:
# check variables.
air_data_subset_dummies.dtypes

satisfaction                          object
Age                                    int64
Flight Distance                        int64
Seat comfort                           int64
Departure/Arrival time convenient      int64
Food and drink                         int64
Gate location                          int64
Inflight wifi service                  int64
Inflight entertainment                 int64
Online support                         int64
Ease of Online booking                 int64
On-board service                       int64
Leg room service                       int64
Baggage handling                       int64
Checkin service                        int64
Cleanliness                            int64
Online boarding                        int64
Departure Delay in Minutes             int64
Arrival Delay in Minutes             float64
Customer Type_Loyal Customer           uint8
Customer Type_disloyal Customer        uint8
Type of Travel_Business travel         uint8
Type of Tr

## **Model building 🤖** 

In [35]:
# Separate the dataset into labels (y) and features (X).
y = air_data_subset_dummies["satisfaction"]
X = air_data_subset_dummies.drop("satisfaction", axis=1)

# Separate into train, validate, test sets.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)
X_tr, X_val, y_tr, y_val = train_test_split(X_train, y_train, test_size = 0.25, random_state = 0)


# Determine set of hyperparameters.
cv_params = {'n_estimators' : [50,100], 
              'max_depth' : [10,50],        
              'min_samples_leaf' : [0.5,1], 
              'min_samples_split' : [0.001, 0.01],
              'max_features' : ["sqrt"], 
              'max_samples' : [.5,.9]}

# Create list of split indices.
split_index = [0 if x in X_val.index else -1 for x in X_train.index]
custom_split = PredefinedSplit(split_index)

# Instantiate model.
rf = RandomForestClassifier(random_state=0)

# Search over specified parameters.
rf_val = GridSearchCV(rf, cv_params, cv=custom_split, refit='f1', n_jobs = -1, verbose = 1)

In [36]:
%%time

# Fit the model
rf_val.fit(X_train, y_train)

Fitting 1 folds for each of 32 candidates, totalling 32 fits


C:\Users\KirubaSagar\anaconda3\lib\site-packages\joblib\externals\loky\process_executor.py:702: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


Wall time: 1min 7s


GridSearchCV(cv=PredefinedSplit(test_fold=array([-1, -1, ..., -1, -1])),
             estimator=RandomForestClassifier(random_state=0), n_jobs=-1,
             param_grid={'max_depth': [10, 50], 'max_features': ['sqrt'],
                         'max_samples': [0.5, 0.9],
                         'min_samples_leaf': [0.5, 1],
                         'min_samples_split': [0.001, 0.01],
                         'n_estimators': [50, 100]},
             refit='f1', verbose=1)

In [49]:
# Obtain optimal parameters.
rf_val.best_params_

{'max_depth': 50,
 'max_features': 'sqrt',
 'max_samples': 0.9,
 'min_samples_leaf': 1,
 'min_samples_split': 0.001,
 'n_estimators': 50}

## **Results and evaluation 🧪** 

In [50]:
%%time

# Use optimal parameters on GridSearchCV.
rf_opt = RandomForestClassifier(n_estimators = 50, max_depth = 50, 
                                min_samples_leaf = 1, min_samples_split = 0.001,
                                max_features="sqrt", max_samples = 0.9, random_state = 0)

# Fit the optimal model.
rf_opt.fit(X_train, y_train)

Wall time: 8.96 s


RandomForestClassifier(max_depth=50, max_features='sqrt', max_samples=0.9,
                       min_samples_split=0.001, n_estimators=50,
                       random_state=0)

In [51]:
# Predict on test set.
y_pred = rf_opt.predict(X_test)

### Obtain performance scores 📑

In [52]:
# Get precision score.
pc_test = precision_score(y_test, y_pred, pos_label = "satisfied")
print("The precision score is {pc:.3f}".format(pc = pc_test))

# Get recall score.
rc_test = recall_score(y_test, y_pred, pos_label = "satisfied")
print("The recall score is {rc:.3f}".format(rc = rc_test))

# Get accuracy score.
ac_test = accuracy_score(y_test, y_pred)
print("The accuracy score is {ac:.3f}".format(ac = ac_test))

# Get F1 score.
f1_test = f1_score(y_test, y_pred, pos_label = "satisfied")
print("The F1 score is {f1:.3f}".format(f1 = f1_test))

The precision score is 0.950
The recall score is 0.945
The accuracy score is 0.942
The F1 score is 0.947


### Evaluate the model 

In [53]:
# Create table of results.

table = pd.DataFrame()
table = table.append({'Model': "Tuned Decision Tree",
                        'F1':  0.945422,
                        'Recall': 0.935863,
                        'Precision': 0.955197,
                        'Accuracy': 0.940864
                      },
                        ignore_index=True
                    )

table = table.append({'Model': "Tuned Random Forest",
                        'F1':  f1_test,
                        'Recall': rc_test,
                        'Precision': pc_test,
                        'Accuracy': ac_test
                      },
                        ignore_index=True
                    )
table

C:\Users\KirubaSagar\AppData\Local\Temp\ipykernel_10176\1213430234.py:4: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  table = table.append({'Model': "Tuned Decision Tree",
C:\Users\KirubaSagar\AppData\Local\Temp\ipykernel_10176\1213430234.py:13: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  table = table.append({'Model': "Tuned Random Forest",


Model        F1    Recall  Precision  Accuracy
0  Tuned Decision Tree  0.945422  0.935863   0.955197  0.940864
1  Tuned Random Forest  0.947306  0.944501   0.950128  0.942450

## **Considerations**

**What summary would you provide to stakeholders?**
* The random forest model predicted satisfaction with more than 94.2% accuracy. The precision is over 95% and the recall is approximately 94.5%. 
* The random forest model outperformed the tuned decision tree with the best hyperparameters in most of the four scores. This indicates that the random forest model may perform better.
* Because stakeholders were interested in learning about the factors that are most important to customer satisfaction, this would be shared based on the tuned random forest. 
* In addition, you would provide details about the precision, recall, accuracy, and F1 scores to support your findings. 